### DISCUSSION TITLE
How to use property setting with command that returns response

### DISCUSSION THREAD

**Discussion by sansanda** [pymeasure/pymeasure#921] [Q&A] [answered]
Hi All.

I want to use the property setting for sending a command to my instrument via GPIB. As simple command could be 'LI0' which power offs  a lamp. Here is my first attemp to resolve it. It runs perfectly. 

```
lamp_on = Instrument.setting(
    "%s",
    "Control lamp on/off of the probe station",
    validator=strict_discrete_set,
    values={True: 'LI1', False: 'LI0'},
    map_values=True
)
```
But, for every command I send to the instrument there is associated a response which I would like to read. How can I solve that using properties?

Best regards,
David.

**Comment by BenediktBurger** [comment, marked as answer]
Hi @sansanda ,

As that problem occurs several times, it is listed in the [solutions for implementation challenges](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/solutions.html#communication-protocol-issues) and was reason for #883 .

Here is a more extensive response:
If you set `check_set_errors=True` for a property, the property will call the `check_set_errors()` method (until #883 the `check_errors` method).
In this method you can read the response and handle it appropriately.

For example:
~~~python
lamp_on = Instrument.setting(
    "LI%s",
    "Control lamp on/off of the probe station",
    validator=strict_discrete_set,
    values={True: '1', False: '0'},
    map_values=True,
    check_set_errors=True,
)

def check_set_errors(self):
    response = self.read()
    if response is fishy:
        log.warning(f"Fishy response '{response}' after setting a value.")
        return [f"Fishy response '{response}'."]
    else:
        return []
~~~

Note that `check_set_errors` (like the get variant and the general one) are supposed to log the errors and to return a list of errors.

If there are more questions, I'm glad to help.

**Reply by sansanda** [reply]
Many thanks @bmoneke.

This check_set_errors will help me to maintain update a variable with the info of the response for the last command executed by the instrument. For me is very helpful. 

David

**Reply by BenediktBurger** [reply]
Glad to help!
